design/legacy-block-storage.md

```diff
@@ -0,0 +1,141 @@
+# Legacy Block Storage with Rook
+
+
+## Overview
+
+### Value of legacy block storage support to Rook
+
+* In order to completely integrate with Kubernetes and utilize Rook's autonomous healing, scaling, and managing features, backend storage systems need to be software-based, distributed (clustered), and capable of being containerized. 
+Not all modern storage systems meet these requirements.
+* Many expensive and proprietary appliance storage silos are still providing services for Enterprise IT infrastructures.
+* Many companies only trust well-certified and time-proven storage solutions for their mission critical applications
+
+For transitioning into existing enterprise IT infrastructures, Rook will have a much easier time with support for legacy block storage systems.
+
+
+### Storage Resource Normalization
+
+The OpenStack Block Storage service (Cinder) provides a standardized architecture to aggregate the features and management of all kinds of block storage systems.
+It is the first service to have a standardized interface that is widely supported by most storage vendors including Dell EMC, IBM, NetApp, Hitachi, HPE, Pure Storage, VMware, etc.
+
+The major functions of the Cinder volume driver include:
+
+* Create/Delete Volume
+* Attach/Detach Volume
+* Extend Volume
+* Create/Delete Snapshot
+* Create Volume from Snapshot
+* Create Volume from Volume (clone)
+* Create Image from Volume
+* Volume Migration (host assisted)
+
+The following are some references for OpenStack Cinder:
+
+* [OpenStack Cinder](https://access.redhat.com/documentation/en/red-hat-enterprise-linux-openstack-platform/version-7/red-hat-enterprise-linux-openstack-platform-7-architecture-guide/chapter-1-components#comp-cinder)
+* [Cinder Driver Support Matrix](https://wiki.openstack.org/wiki/CinderSupportMatrix)
+
+In this project, we propose to leverage the open source project [OpenStack Cinder](https://github.com/openstack/cinder) as the unified storage connector that integrates legacy block storage with Rook.
+Through this implementation, we can quickly extend Rook's storage backend support to most modern block storage solutions that are already used in production today.
+
+
+## Architecture
+
+![legacy-block-storage-architecture-diagram](legacy-block-storage-architecture.png)
+
+### Operator
+
+* start and monitor the Cinder service pods
+* watch for state changes and apply the changes
+* monitor the health of external storage systems
+* manage custom resource definitions (CRDs) for pools
+
+
+### Agent
+
+* deployed on each Kubernetes nodes
+* handle attaching/detaching, mounting/unmounting volumes and formatting the file system
+
+
+## Details
+
+### lbs (legacy block storage) cluster CRD example
+
+```yaml
+apiVersion: lbs.rook.io/v1alpha1
+kind: Cluster
+metadata:
+  name: rook-lbs
+  namespace: rook-lbs
+spec:
+  cinderVersion:
+    image: openstack/cinder:v13.0.2-4
+  dataDirHostPath: /var/lib/rook
```

Comment:
`dataDirHostPath` is for persisting data to the host system. For ceph this is necessary because mons and osds need to persist their highly stateful data, I don't see anything about the cinder config that suggests data persistence should be necessary, though I may very well be wrong. Persisting data to the host will make things a lot more difficult than if k8s can just hold the data. The lbs implementation will have to support the case where the host on which data is persisted dies and the data stored on the host needs to be regenerated regardless, and I think that unless this process is incredibly time-consuming, this should be the main behavior which will eliminate the need for host persistence.

As an example in Ceph of non-persistent data, the `/etc/ceph/ceph.conf` file is generated by each daemon pod's `config-init` container and stored in an EmptyDir to be later used by the daemon container. In the best case, I think this is the model that can be followed for the lbs (cinder) implementation. It looks like all of the cinder driver configuration information is captured by the crd. (for example, the `san_ip` and `netapp_server_hostname` are stored by k8s in the `lbs.Cluster` resource. As a note: although Ceph's config file is generated in the `config-init` container, my experience has been that this is unnecessarily complicated, and I would like to see the `config-init` behavior moved to the Ceph operator, which I would suggest for any Rook operator implementation unless it is impossible.

Please include a brief description of the data Cinder needs to persist (if any) and how that data is created. Please also describe how the lbs operator will bootstrap the cinder pods. If possible, configuring cinder using only commandline arguments and not using a config file will be easier to implement and maintain. If a config file is needed, my thought would be to use k8s secrets for storing cinder config files which can then be mounted to pods at a certain directory in containers for cinder to use; however, this likely isn't the only way of doing such a thing.